src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,163 @@
+        conn.transaction(|conn| {
+            for token in chunk {
+                // Send notification.
+                let user = User::find(conn, token.user_id)?;
+                let Some(recipient) = user.email(conn)? else {
+                    return Err(anyhow!("No address found"));
+                };
+                let email = ExpiryNotificationEmail {
+                    name: user.gh_login.clone(),
+                    token_name: token.name.clone(),
+                };
+                match emails.send(&recipient, email) {
+                    Ok(_) => {
+                        // Update the token to prevent duplicate notifications.
+                        diesel::update(token)
+                            .set(
+                                crate::schema::api_tokens::expiry_notification_at
+                                    .eq(now.nullable()),
+                            )
+                            .execute(conn)?;
+                    }
+                    Err(e) => {
+                        error!("Failed to send email: {:?} to {}", e, recipient);
+                    }
+                }
+            }
+            Ok::<_, anyhow::Error>(())
+        })?;
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+struct ExpiryNotificationEmail {
+    name: String,
+    token_name: String,
+}
+
+impl Email for ExpiryNotificationEmail {
+    const SUBJECT: &'static str = "Your token is about to expire";
+
+    fn body(&self) -> String {
+        format!(
+            r#"Hi @{},
+
+We noticed your token "{}" will expire in about {} days.
+
+If this token is still needed, visit https://crates.io/settings/tokens/new to generate a new one.
+
+Thanks,
+The crates.io team"#,
+            self.name, self.token_name, EXPIRY_THRESHOLD
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        models::token::ApiToken, schema::api_tokens, test_util::test_db_connection,
+        typosquat::test_util::Faker, util::token::PlainToken,
+    };
+    use diesel::dsl::IntervalDsl;
+    use diesel::{QueryDsl, SelectableHelper};
+    use lettre::Address;
+
+    #[tokio::test]
+    async fn test_expiry_notification() -> anyhow::Result<()> {
+        let emails = Emails::new_in_memory();
+        let (_test_db, mut conn) = test_db_connection();
+        let mut faker = Faker::new();
+
+        // Set up a user and a token that is about to expire.
+        let user = faker.user(&mut conn, "a", Some("testuser@test.com".to_owned()))?;
+        let token = PlainToken::generate();
+
+        let token: ApiToken = diesel::insert_into(api_tokens::table)
+            .values((
+                api_tokens::user_id.eq(user.id),
+                api_tokens::name.eq("test_token"),
+                api_tokens::token.eq(token.hashed()),
+                api_tokens::expired_at.eq(now.nullable() + 2.days()),
+            ))
+            .returning(ApiToken::as_returning())
+            .get_result(&mut conn)?;
+
```

Comment:
It would be helpful if we also checked that non-expired tokens are not affected.

Reply:
Good suggestion! Added.